core/tests/test_utils.py

```diff
@@ -760,64 +1005,264 @@ class TestBase(unittest.TestCase):
-        script that calls the test.
-        """
-        # We are using the b' prefix as all the stdouts are in bytes.
-        python_utils.PRINT(
-            b'%s%s' % (LOG_LINE_PREFIX, python_utils.convert_to_bytes(line)))
+        policy = (
+            datastore_services.make_instantaneous_global_consistency_policy())
+        self.testbed.init_datastore_v3_stub(consistency_policy=policy)
+
+        # The root path tells the testbed where to find the queue.yaml file.
+        self.testbed.init_taskqueue_stub(root_path=os.getcwd())
+        self._taskqueue_stub = (
+            self.testbed.get_stub(testbed.TASKQUEUE_SERVICE_NAME))
+
+        # Set up the app to be tested.
+        self.testapp = webtest.TestApp(main.app)
+        self.taskqueue_testapp = webtest.TestApp(main_taskqueue.app)
+        self.mail_testapp = webtest.TestApp(main_mail.app)
+
+        self.signup_superadmin_user()
+
+    def tearDown(self):
+        datastore_services.delete_multi(
+            datastore_services.query_everything().iter(keys_only=True))
+        self.testbed.deactivate()
 
     def login(self, email, is_super_admin=False):
         """Sets the environment variables to simulate a login.
 
         Args:
             email: str. The email of the user who is to be logged in.
             is_super_admin: bool. Whether the user is a super admin.
-       """
-        os.environ['USER_EMAIL'] = email
-        os.environ['USER_ID'] = self.get_gae_id_from_email(email)
-        os.environ['USER_IS_ADMIN'] = '1' if is_super_admin else '0'
+        """
+        self.testbed.setup_env(
+            overwrite=True, user_email=email,
+            user_id=self.get_gae_id_from_email(email),
+            user_is_admin=('1' if is_super_admin else '0'))
 
     def logout(self):
         """Simulates a logout by resetting the environment variables."""
-        os.environ['USER_EMAIL'] = ''
-        os.environ['USER_ID'] = ''
-        os.environ['USER_IS_ADMIN'] = '0'
+        self.testbed.setup_env(
+            overwrite=True, user_email='', user_id='', user_is_admin='0')
 
-    def shortDescription(self):
-        """Additional information logged during unit test invocation."""
-        # Suppress default logging of docstrings.
-        return None
+    @contextlib.contextmanager
+    def login_context(self, email, is_super_admin=False):
+        """Log in with the given email under the context of a 'with' statement.
+
+        Args:
+            email: str. An email associated to a user account.
+            is_super_admin: bool. Whether the user is a super admin.
+
+        Yields:
+            str. The id of the user associated to the given email, who is now
+            'logged in'.
+        """
+        self.login(email, is_super_admin=is_super_admin)
+        try:
+            yield self.get_user_id_from_email(email)
+        finally:
+            self.logout()
+
+    def superadmin_context(self):
```

Comment:
Nope the annotation isn't required here. Functions decorated with contextlib.contextmanager return generator-like objects. This method is taking that decorated generated-like value and returning it directly, so there's no need for it to do the same work again.

Because of this, I decided to make the names consistent. Their usage-pattern and even return type are identical, so it'd be less consistent for their names to be different.

Reply:
Got it, thanks. That said -- while it isn't required to use that annotation, is it possible to use it? If not, then maybe change to login_as_superadmin_context() instead.

Reply:
Done.